projects/npm-tools/packages/npm-scripts/src/sass/build.js

```diff
@@ -0,0 +1,124 @@
+/**
+ * SPDX-FileCopyrightText: © 2020 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+const fs = require('fs');
+const globby = require('globby');
+const r2 = require('liferay-r2');
+const path = require('path');
+const sass = require('sass');
+
+const log = require('../utils/log');
+
+const SASS_EXCLUDE = [
+	'**/_*.scss',
+	'**/build/**',
+	'**/classes/**',
+
+	// Copy from CSSBuilderArgs.java
+
+	'**/_diffs/**',
+	'**/.sass-cache*/**',
+	'**/.sass_cache_*/**',
+	'**/_sass_cache_*/**',
+	'**/_styled/**',
+	'**/_unstyled/**',
+	'**/css/aui/**',
+	'**/css/clay/**',
+	'**/tmp/**',
+];
+
+const getClayPaths = () => {
+	try {
+		return require('@clayui/css').includePaths;
+	} catch (e) {
+		return [];
+	}
+};
+
+const clayIncludePaths = getClayPaths();
+
+async function collectSassFiles(baseDir, excludes = []) {
+	const excludePaths = [
+		...SASS_EXCLUDE.map((exclude) => path.join(baseDir, exclude)),
+		...excludes.map((exclude) => path.join(baseDir, exclude)),
+	];
+
+	return await globby([
+		path.join(baseDir, '**/*.scss'),
+		...excludePaths.map((exclude) => `!${exclude}`),
+	]);
+}
+
+function buildSass(file, output) {
+	const newFileName = file
+		.split('/')
+		[file.split('/').length - 1].replace('.scss', '.css');
+
+	const outputFile = path.join(output, newFileName);
+
+	const result = sass.renderSync({
+		file,
+		includePaths: clayIncludePaths,
+		outFile: outputFile,
+		sourceMap: true,
+	});
+
+	return [outputFile, result.css, result.map];
+}
+
+function copyCss(filePath, newDirectory) {
+	if (!fs.existsSync(newDirectory)) {
+		fs.mkdirSync(newDirectory, {recursive: true});
+	}
+
+	fs.copyFileSync(filePath, path.join(newDirectory, path.basename(filePath)));
+}
+
+const SASS_DIR = '.sass-cache';
+
+async function main(baseDir, {excludes, outputDir = '', rtl = false}) {
+	const files = await collectSassFiles(baseDir, excludes);
+
+	if (!files.length) {
+		log(`BUILD CSS: No files found.`);
+
+		return;
+	}
+
+	for (let i = 0; i < files.length; i++) {
+		const file = files[i];
+
+		const baseBuildDirectory = path.join(
+			baseDir,
+			outputDir,
+			path.parse(file).dir.replace(baseDir, '')
```

Comment:
The last portion is used to append any extra nested file structure found in the glob. Using `path.relative(baseDir, File)` is a good idea, but it will still return the filename and we just want the directory. So we can do `path.dirname(path.relative(baseDir, file))`

Reply:
Thanks for the explanation. This stuff is hard to read. Your `path.dirname(path.relative(baseDir, file))` looks good to me.